### ISSUE TITLE
[Bug]: Can't retrieve role claims for a service user

### ISSUE THREAD

**Issue by sp132** [zitadel/zitadel#6662]
### Preflight Checklist

- [X] I could not find a solution in the documentation, the existing issues or discussions
- [ ] I have joined the [ZITADEL chat](https://zitadel.com/chat)

### Environment

Self-hosted

### Version

2.37.1

### Describe the problem caused by this bug

`/oidc/v1/userinfo` response doesn't contain role claims if called with an access token obtained via client credentials grant.
Calling `/auth/v1/permissions/me/_search` with the same token yields an error:

```json
{
  "code":5,
  "message":"User grant not found (QUERY-wIPkA)",
  "details":[
    {
      "@type":"type.googleapis.com/zitadel.v1.ErrorDetail",
      "id":"QUERY-wIPkA",
      "message":"User grant not found"
    }
  ]
}
```

### To reproduce

1. Create a project
2. Enable "Assert roles on authentication" in project settings
3. Create a role in project
4. Create a service user
5. Generate a client secret for a service user
6. Grant service user a role inside the project
7. Obtain an access token for the service user
```bash
curl -X POST 'https://{ZITADEL_DOMAIN}/oauth/v2/token' \
-H 'Authorization: Basic {AUTH}'  \ 
-H 'Content-Type: application/x-www-form-urlencoded'  \
--data 'scope=openid urn:zitadel:iam:org:project:id:zitadel:aud urn:zitadel:iam:org:project:id:{PROJECT_ID}:aud urn:zitadel:iam:org:project:role:{ROLE_NAME}' \
--data 'grant_type=client_credentials'
```
8. Check userinfo
```bash
curl -H 'Authorization: Bearer {ACCESS_TOKEN}' 'https://{ZITADEL_DOMAIN}/oidc/v1/userinfo'
```
Response contains only `sub` claim
9. Check permissions
```bash
curl -H 'Authorization: Bearer {ACCESS_TOKEN}' 'https://{ZITADEL_DOMAIN}/auth/v1/permissions/me/_search'
```
Observe an error

### Expected behavior

`/oidc/v1/userinfo` contains role claims

**Comment by sp132** [comment]
Could anyone confirm whether this is really a bug or I'm missing some configuration pieces?

**Comment by hifabienne** [comment]
Hei @sp132 
I checked this right now instead of using the specific rolekey, I used the roles scope to get all the roles back. And this worked for me:
curl --request POST \                                                                                                                                                                                            
  --url https://{CUSTOM-DOMAIN}/oauth/v2/token \
  --header 'Content-Type: application/x-www-form-urlencoded' \
  --header 'Authorization: Basic {AUTH}' \
  --data grant_type=client_credentials \
  --data scope='openid profile email urn:zitadel:iam:org:projects:roles urn:zitadel:iam:org:project:id:{PROJECT ID}:aud'

**Comment by sp132** [comment]
Hi @hifabienne

Thanks, using `urn:zitadel:iam:org:projects:roles` instead of `urn:zitadel:iam:org:project:role:{rolekey}` helped.

Does this mean I should avoid using `urn:zitadel:iam:org:project:role:{rolekey}` at all?

**Comment by hifabienne** [comment]
Hei @sp132 
I think actually the other scope should work as well. But could not yet figure out how.
@stebenz @livio-a Do you know if there is a bug or if we do something wrong?

**Comment by livio-a** [comment]
I was finally able to have a look at this and it seems that there is a bug, at least for service accounts using the single role scope. I could narrow it down, but need to check how to fix it.

hints for later:
- applicationID and roleAudience are empty since it not going through a code flow:
https://github.com/zitadel/zitadel/blob/main/internal/api/oidc/client.go#L812-L814

**Comment by muhlemmer** [comment]
@livio-a  I just had a look if the userinfo optimization from #6910 solved this, but it didn't.

Basically the problem is that the token does not carry a client_id, because there was never a auth flow. We retrieve the project by client_id only if it was present in the token. The custom audience scopes are basically ignored here.

It would probably work fine with introspection, as there we take the project from the authenticated client.

I will need to think a bit more about a solution later.

**Comment by forsak3n** [comment]
for what it's worth, introspection doesn't work with service user credentials (2.43.0 self-hosted). It always returns `{"active": false}`.

**Comment by muhlemmer** [comment]
Note that on the linked fix branch, `/auth/v1/permissions/me/_search` works fine:

```json
{
  "details": {
    "totalResult": "1",
    "viewTimestamp": "2024-04-13T14:46:25.446177Z"
  },
  "result": [
    {
      "orgId": "262592437451579393",
      "projectId": "262601670708453377",
      "userId": "262601591050231809",
      "roles": [
        "bar",
        "foo"
      ],
      "orgName": "ZITADEL",
      "grantId": "262601734394765313",
      "details": {
        "sequence": "1",
        "creationDate": "2024-04-13T14:33:50.961813Z",
        "changeDate": "2024-04-13T14:33:50.961813Z",
        "resourceOwner": "262592437451579393"
      },
      "orgDomain": "zitadel.localhost",
      "projectName": "test",
      "roleKeys": [
        "bar",
        "foo"
      ],
      "userType": "TYPE_MACHINE"
    }
  ]
}
```

I haven't  touched any code in the API, but perhaps this was solved sometime between 2.37 and now.  I will not investigate further and assume the fix in #7763  is sufficient to close this issue.
